New exception handling diagnostic fixes

janvorli opened the pull request:
This change makes the new exception handling pass all the mdbg tests (tested on Windows x64, Linux x64 and Linux arm64) except for the Exceptions.ReflectionTest that will require a change in the test. 
Here is a list of changes:
* Adds debugger and profiler events in the right places and order
* Adds debugger exception interception implementation
* Ensures that the new managed exception handling frames are not visible to the debugger stack trace
* Fixes some missing frames on stack traces (coming from `HelperMethodFrame`)
* Moves first chance exception handling to the native code. It is necessary to ensure that the exception stack trace contains frames from the explicit frames that have `MethodDesc` at the time the event fires.
* Makes some refactoring around the `ExInfo` - the `CONTEXT` and `REGDISPLAY` are now in the `ExInfo`, the `ExInfo` now also contains the original exception pointers.
* Fixes one case when an unhandled exception at the boundary of managed and native code was not handled correctly.
* Changes the way an unhandled exception in exception filter is swallowed. The stack frame iterator was having issues with the fact that the exception "leaked" from the actual filter and was caught by another filter in the managed EH. Now the exception is always swallowed at the filter.

janvorli committed:
New exception handling diagnostic fixes

This change
* Adds debugger and profiler events in the right places and order
* Adds debugger exception interception implementation
* Ensures that the new managed exceptino handling frames are not visible
  to the debugger stack trace
* Fixes some missing frames on stack traces (coming from
  HelperMethodFrame)
* Moves first chance exception handling to the native code. It is
  necessary to ensure that the exception stack trace contains frames
  from the explicit frames that have MethodDesc at the time the event
  fires.
* Makes some refactoring around the ExInfo - the CONTEXT and REGDISPLAY
  are now in the ExInfo, the ExInfo contains the original exception pointers.
* Fixes one case when an UnhandledException at the boundary of managed
  and native code was not handled correctly.
* Changes the way an unhandled exception in exception filter is
  swallowed. The stack frame iterator was having issues with the fact
  that the exception "leaked" from the actual filter and was caught by
  another filter in the managed EH. Now the exception is always
  swallowed at the filter.

tommcdon commented:
@mikem8361

jkotas commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -957,19 +1023,23 @@ private static void DebugVerifyHandlingFrame(UIntPtr handlingFrameSP)
                     byte* pFilterFunclet = ehClause._filterAddress;
 
                     bool shouldInvokeHandler = false;
+#if NATIVEAOT
                     try
                     {
+#endif
                         shouldInvokeHandler =
 #if NATIVEAOT
                             InternalCalls.RhpCallFilterFunclet(exception, pFilterFunclet, frameIter.RegisterSet);
 #else
                             InternalCalls.RhpCallFilterFunclet(ObjectHandleOnStack.Create(ref exception), pFilterFunclet, frameIter.RegisterSet);
 #endif
+#if NATIVEAOT
                     }
                     catch when (true)
                     {
                         // Prevent leaking any exception from the filter funclet
                     }
+#endif
```
```suggestion
#if NATIVEAOT
                    try
                    {
                        shouldInvokeHandler = InternalCalls.RhpCallFilterFunclet(exception, pFilterFunclet, frameIter.RegisterSet);
                    }
                    catch when (true)
                    {
                        // Prevent leaking any exception from the filter funclet
                    }
#else // NATIVEAOT
                    InternalCalls.RhpCallFilterFunclet(ObjectHandleOnStack.Create(ref exception), pFilterFunclet, frameIter.RegisterSet);
#endif // NATIVEAOT
```

janvorli replied:
Right, it looks better. I will update it this way.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/AppContext.cs:
```diff
@@ -82,12 +82,7 @@ public static void SetData(string name, object? data)
-            FirstChanceException?.Invoke(AppDomain.CurrentDomain, new FirstChanceExceptionEventArgs((Exception)e));
-        }
-#endif
+        internal static event EventHandler? ProcessExit;
```
Build break - incorrectly resolved merge conflict?

janvorli replied:
Yes, fixed

janvorli committed:
Fix bad merge

janvorli committed:
Fix win x86 build break

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -2300,6 +2306,46 @@ bool ExceptionTracker::HandleNestedExceptionEscape(StackFrame sf, bool fIsFirstP
+
+    BOOL fDeliveredFirstChanceNotification = FALSE;
+
+    // <TODO>
+    // Remove this once SIS is fully enabled.
+    // </TODO>
+    extern bool g_EnableSIS;
```
Delete this from here and everywhere else? It seems to be unconditionally enabled, so this TODO is no longer relevant.

janvorli replied:
Deleted

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -2300,6 +2306,46 @@ bool ExceptionTracker::HandleNestedExceptionEscape(StackFrame sf, bool fIsFirstP
+        // We recognize stubs as Frames with a M2U transition type. The debugger's stackwalker also
+        // recognizes these frames and publishes ICorDebugInternalFrames in the stackwalk. It's
+        // important to use pFrame as the stack address so that the Exception callback matches up
+        // w/ the ICorDebugInternlFrame stack range.
```
```suggestion
        // w/ the ICorDebugInternalFrame stack range.
```
Pre-existing typo (please fix other instances too)

janvorli replied:
Fixed

janvorli committed:
Reflect PR feedback

janvorli committed:
Fix Watson not being triggered on unhandled exception

janvorli committed:
Modify the unhandled exception processing

Slight modification of the code enables second pass for unhandled
exceptions. That makes finallys get called the same way as without the
new exception handling.

tommcdon reviewed (approved):
LGTM, thanks!

janvorli committed:
Fix accidental edit

janvorli commented:
The windows arm64 nativeaot leg failure is a known issue
